File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.py
def solution(dartResult):
    numberarr = ['0','1','2','3','4','5','6','7','8','9']
    pointarr = []
    cnt = 0
    for i in range(len(dartResult)) : 
        now = dartResult[i]
        
        if now == '1' and dartResult[i+1] == '0' :
            cnt = cnt + 10
            continue
        
        elif now in numberarr : 
            cnt = cnt + int(now)
            continue
            
        if now == 'S' :
            pointarr.append(cnt)
        elif now == 'D' :
            pointarr.append(cnt**2)
        elif now == 'T':
            pointarr.append(cnt**3)
        
        elif now == '#' :
            pointarr[-1] = pointarr[-1]* (-1)
        elif now == '*':
            if len(pointarr) == 1:
                pointarr[-1] = pointarr[-1] *2 
            else :
                pointarr[-1] = pointarr[-1]*2
                pointarr[-2] = pointarr[-2]*2
        
        
        cnt = 0
        
    return sum(pointarr)     
        
 

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.java
import java.util.*;
class Solution {
    
    public char [] point = new char [] {'1','2','3','4','5','6','7','8','9','0'};
    public char [] bonus = new char[] {'S','D','T'}; //idx+1 제곱
    public char [] option = new char[] {'*','#'} ;
    public int answer=  0 ;
    public int before = 0 ;
    public int solution(String dartResult) {
        StringBuilder sb=  new StringBuilder();
        for (int i = 0; i<dartResult.length(); i++)
        {
            char now = dartResult.charAt(i);
            sb.append(now);
            if (findidx(now, bonus) != -1 && i<dartResult.length()-1 && findidx(dartResult.charAt(i+1),option) == -1) {
            int nowval = calculate(sb.toString());
            answer += nowval;
            before = nowval;
            System.out.println("nowval : " + nowval + " ans : " + answer);
            sb.setLength(0);
            }
        // 옵션(*,#)도 계산 시작
        else if (findidx(now, option) != -1) {
            int nowval = calculate(sb.toString());
            answer += nowval;
            before = nowval;
            System.out.println("nowval : " + nowval + " ans : " + answer);
            sb.setLength(0);
        }
        }
        
        int nowval = calculate(sb.toString());
                    answer += nowval;
                    before = nowval;
                    System.out.println("nowval : "+nowval+ " ans : "+ answer);
        
        
        return answer;
    }
    
    public int calculate(String ans)
    {
        int val = 0 ;
        StringBuilder num = new StringBuilder();
        for (int i = 0 ; i<ans.length() ; i++)
        {
            char now = ans.charAt(i);
           if (findidx(now,point) != -1)
           {
               num.append(now);
           } else if (findidx(now,bonus) != -1)
           {
                val = Integer.parseInt(num.toString());  
                val = (int) Math.pow(val, findidx(now, bonus) + 1);
                num.setLength(0);
           } else if (findidx(now,option)!=-1)
           {
              if (now == '*')
              {
                  answer += before;
                  val *= 2;
              } else {
                  val = (0-val);
              }
           }
        }
        
        if (num.length() > 0) {
            val = Integer.parseInt(num.toString());
        }

        
        return val;
    }
    
    public int findidx (char val , char [] target)
    {
        for (int i = 0 ; i<target.length; i++)
        {
            
        if (target[i] == val)
        {
            return i;
        }
        }
        return -1;
    }
}